### DISCUSSION TITLE
Writing a test that uses coroutines interop?

### DISCUSSION THREAD

**Discussion by plusmobileapps** [badoo/Reaktive#657] [Q&A] [answered]
Has anybody ever written a test which uses the coroutines interop? I was trying to write one with the following being mocked and the class under test

```kotlin
interface FirebaseAuth {
    val authState: StateFlow<FirebaseAuthState>
}

internal class AuthenticationRepositoryImpl(private val auth: FirebaseAuth) :
    AuthenticationRepository {

    override val authState: Observable<AuthState> =
        auth.authState.asObservable()
            .subscribeOn(ioScheduler)
            .map { state ->
                when (state) {
                    is FirebaseAuthState.Authenticated -> AuthState.Authenticated(
                        state.firebaseUser.toWolfPackUser()
                    )
                    FirebaseAuthState.Unauthenticated -> AuthState.Unauthenticated
                }
            }
}
```

The following test seems to work on android, but hangs on iOS and struggling to figure out what's causing the test to hang? It looks like the `asObservable()` is technically launching a coroutine so I thought the `runTest { }` was maybe needed from the [kotlinx-coroutines-test](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-test/README.md), but then why would this test still pass on Android without it? 

```kotlin
internal class AuthenticationRepositoryTest : TestsWithMocks() {

    override fun setUpMocks() = injectMocks(mocker)

    @Mock
    lateinit var firebaseAuth: FirebaseAuth

    private val repository by withMocks { AuthenticationRepositoryImpl(firebaseAuth) }

    @BeforeTest
    fun before() {
        overrideSchedulers(main = { TestScheduler() }, io = { TestScheduler() })
    }

    @Test
    fun authState_IsAuthenticated() {
        val firebaseUser: FirebaseUser =
            fakeFirebaseUser().copy(uid = "some id", email = "some email", displayName = "my name")
        every { firebaseAuth.authState } returns MutableStateFlow(
            FirebaseAuthState.Authenticated(firebaseUser)
        )
        repository.authState
            .test()
            .assertValue(
                AuthState.Authenticated(
                    WolfPackUser(
                        userId = firebaseUser.uid,
                        email = firebaseUser.email,
                        displayName = firebaseUser.displayName
                    )
                )
            )
    }
```

The test is using [MocKmp](https://github.com/Kodein-Framework/MocKMP) for generating the mocks in the test by the way which works great when the class under test is all reaktive using a similar approach.

**Comment by arkivanov** [comment, marked as answer]
Do you use a stable version of coroutines or `-native-mt`? What is the purpose of `subscribeOn(ioScheduler)` in this particular case?

**Reply by plusmobileapps** [reply]
I was using the multi threaded version with the new memory model enabled in 1.6.0 and the subscribeOn was needed otherwise the iOS app would only have a black screen which I assume was because it was locking up the main thread?

**Reply by arkivanov** [reply]
I see multiple issues here. 

First of all, from my point of view - the native-mt coroutines shouldn't be required with the new memory model? As per [this comment](https://github.com/Kotlin/kotlinx.coroutines/issues/462#issuecomment-999557659). 

Second, from my point of view, the thread switching should be performed in the place where it is actually required. In this case, I would switch threads somewhere in `FirebaseAuth` implementation. You can add `observeOn(mainScheduler)` if you need emissions on the main thread, but looks like you don't.

Please let me know if you still have issues after applying suggestions, and I will check the code locally.

**Reply by plusmobileapps** [reply]
Sorry I confuse the`-native-mt` with just the new memory model for multithreaded, so I was just using the the latests version of coroutines with the new memory model. 

For the second point, I ended up making a reaktive version of `FirebaseAuth` which takes care of subscribing on the `ioScheduler` so I could remove all of those calls from the repository. Having converted that over to reaktive now, the testing isn't an issue since the repository is all reaktive. Although still a little confused on how to write a test that has coroutines in it with reaktive. Not sure if there are any examples of this anywhere?

**Reply by arkivanov** [reply]
Thanks for the update! I will try to write such a test, and provide updates here.

**Reply by arkivanov** [reply]
Btw, if you are using the stable version of the Reaktive's `coroutines-interop` module, then that's might be the issue. It collects the Flow from `runBlocking`, and so it hangs. You can try the `-nmtc` version. Please see the readme [section](https://github.com/badoo/Reaktive#coroutines-interop-based-on-stable-kotlinxcoroutines).

**Reply by arkivanov** [reply]
@plusmobileapps This discussion uncovered an issue in Reaktive under the new Kotlin/Native memory model. It freezes things when changing schedulers, and frozen objects still can't be mutated under the new memory model. In the provided`MutableStateFlow` gets frozen and then crashes when collected. Filed #658.

**Reply by plusmobileapps** [reply]
I think it might have been because I was using the stable version and I also had the grade property on for disabling freezing. But that was intended only for libraries that haven't migrated yet and are still freezing. 

https://github.com/JetBrains/kotlin/blob/master/kotlin-native/NEW_MM.md

Thank you for investigating and filing an issue.

**Reply by arkivanov** [reply]
Wow! Thanks for mentioning the way to disable freezing! In this case we don't need to do any changes in the library, since users may just disable freezing entirely. I will add the information to the readme instead.

Now I can confirm, collecting a `Flow` via `Flow.asObservable()` operator hangs, if subscribed on the main thread. This is because the stable version of the `coroutines-interop` module uses `runBlocking` internally. Switching to the `-nmtc` version solves the issue and the test passes.

Here is the test:

```kotlin
class AuthenticationRepositoryTest {

    @BeforeTest
    fun before() {
        overrideSchedulers(main = { TestScheduler() }, io = { TestScheduler() })
    }

    @Test
    fun foo() {
        val stateFlow = MutableStateFlow(FirebaseAuthState(x = 1))
        val auth = TestFirebaseAuth(stateFlow)
        val impl = AuthenticationRepositoryImpl(auth)

        impl.authState.test().assertValue(AuthState(x = 1))
    }
}

class TestFirebaseAuth(
    override val authState: StateFlow<FirebaseAuthState>
) : FirebaseAuth

class FirebaseAuthState(val x: Int)

interface FirebaseAuth {
    val authState: StateFlow<FirebaseAuthState>
}

data class AuthState(val x: Int)

internal class AuthenticationRepositoryImpl(auth: FirebaseAuth) {

    val authState: Observable<AuthState> =
        auth.authState.asObservable()
            .subscribeOn(ioScheduler)
            .map { state -> AuthState(x = state.x) }
}
```
